crates/host/src/wasmbus/providers/mod.rs

```diff
@@ -1,2 +1,563 @@
+                                        );
+                                    }
+                                },
+                                (Ok(HealthCheckResponse { healthy: false, ..}), true) => {
+                                    trace!(?provider_id, "provider health check failed");
+                                    previous_healthy = false;
+                                    if let Err(e) = event::publish(
+                                        &event_builder,
+                                        &ctl_nats,
+                                        &lattice,
+                                        "health_check_failed",
+                                        event::provider_health_check(
+                                            &host_id,
+                                            &provider_id,
+                                        )
+                                    ).await {
+                                        warn!(
+                                            ?e,
+                                            ?provider_id,
+                                            "failed to publish provider health check failed event",
+                                        );
+                                    }
+                                }
+                                // If the provider health status didn't change, we simply publish a health check status event
+                                (Ok(_), _) => {
+                                    if let Err(e) = event::publish(
+                                        &event_builder,
+                                        &ctl_nats,
+                                        &lattice,
+                                        "health_check_status",
+                                        event::provider_health_check(
+                                            &host_id,
+                                            &provider_id,
+                                        )
+                                    ).await {
+                                        warn!(
+                                            ?e,
+                                            ?provider_id,
+                                            "failed to publish provider health check status event",
+                                        );
+                                    }
+                                },
+                                _ => warn!(
+                                    ?provider_id,
+                                    "failed to deserialize provider health check response"
+                                ),
+                            }
+                        }
+                        else {
+                            warn!(?provider_id, "failed to request provider health, retrying in 30 seconds");
+                        }
+                }
+                true = async { shutdown.load(Ordering::Relaxed) } => {
```

Comment:
I feel like this is basically going to spinloop here because the future will return almost immediately and then enter the next loop. If the shutdown is going to be used this way, maybe it would be better to use `tokio::sync::watch` and then use the `changed()` method as the future to wait on here. In other places where you want to return the current value syncronously, you could use `borrow()` method

Reply:
Upon further reflection, I don't even need to have this branch in the `select!` statement because shutting down aborts all of the tasks anyways 😄

Reply:
In the case where the provider fails to restart, we will still have the health check tasks available now to report that it is unhealthy as well.